### DISCUSSION TITLE
Outdated PRs Opened for Lock File Maintenance

### DISCUSSION THREAD

**Discussion by Kurt-von-Laven** [renovatebot/renovate#18794] [Request Help]
Renovate opened ScribeMD/.github#119, which at the time was about 9 commits behind main and contained no changes that were not already present on main. I suspect this may be the same issue described in #13126. When I merged that PR, Renovate immediately opened the correct PR, ScribeMD/.github#120. When I merged that PR, Renovate immediately opened the vacuous PR, ScribeMD/.github#121, because of python-poetry/poetry#6201. When I merged that PR, Renovate immediately opened the duplicate PR, ScribeMD/.github#122. [This](https://github.com/ScribeMD/.github/blob/630879258d991846fc5e12882cb4703f9ce8580d/default.json) is the Renovate config we were using during these happenings.

**Comment by rarkins** [comment]
Was the commit made at the same time as the PR was created, or was it made e.g. days earlier?

**Reply by Kurt-von-Laven** [reply]
It was made [about 6 days earlier](https://github.com/ScribeMD/.github/pull/119/commits).

**Reply by rarkins** [reply]
Do you know what led to it being created earlier but only PR'd later? e.g. would it be `"prConcurrentLimit": 1,`?

**Reply by Kurt-von-Laven** [reply]
I am not sure. I am considering overriding the default schedule to `["at any time"]`, which would almost surely address the issue, but also quite possibly bury a bug. I can try removing `"prConcurrentLimit"` in a test repo.

**Reply by Kurt-von-Laven** [reply]
I just realized that ScribeMD/.github#119 was itself a duplicate of ScribeMD/.github#103, which was created by the default lock file maintenance schedule the preceding week. When I merged the latter PR, the branch was never deleted from the fork. This has been the case for all lock file maintenance PRs and no others, possibly because the former are [immortal](https://docs.renovatebot.com/key-concepts/pull-requests/#immortal-prs)? It sounds like you already have a plan to eliminate immortal PRs, so it seems unlikely that I will bury a bug long-term by setting the schedule to `["at any time"]`. Does that sound plausible?

**Reply by Kurt-von-Laven** [reply]
It does seem to me like the opening of duplicate and vacuous (meaning essentially whitespace-only in this case) PRs are both bugs. Ideally, Forking Renovate would reliably sync the fork before creating a new branch/PR when triggered by a merge. `poetry update` does not create any changes when run on our repositories currently, so I am guessing Renovate is doing something a little more sophisticated in order to generate diffs, such as ScribeMD/template-repo#40, that only add two spaces of indentation. Another approach might be simply to never create a branch containing whitespace-only changes.

**Reply by Kurt-von-Laven** [reply]
[Setting `schedule` to `["at any time"]`](https://github.com/ScribeMD/.github/blob/204ee403ef99a02120b80f12a2f3415e79760e6f/default.json#L28) resulted in the creation of ScribeMD/.github#132. I was hoping that branch/PR creation would have been controlled by the Dependency Dashboard, but it was not. Is there some means by which I can control lock file maintenance via the Dependency Dashboard as a workaround?

**Reply by viceice** [reply]
checkout renovate config in this repo. 😉

**Reply by viceice** [reply]
and the corresponding preset 🙃

**Reply by Kurt-von-Laven** [reply]
Did you intend to include a link?

**Reply by Kurt-von-Laven** [reply]
@viceice, you may have had something better in mind, but I set the schedule to a non-existent date (e.g., "on the 29th day of February in 1970") to effectively disable it. I do think there are some genuine bugs here though as mentioned previously.

**Reply by viceice** [reply]
ok, we're using `:maintainLockFilesWeekly` preset. you can enable dashboard approval via package rule

you need `matchUpdateTypes: ["lockFileMaintenance"]`.

**Reply by viceice** [reply]
https://github.com/renovatebot/.github/blob/861063b66dd748c20eb2ef588118289920f0569e/default.json#L36

**Reply by rarkins** [reply]
Schedule controls branch creation only. If the branch exists then it changes nothing

**Reply by Kurt-von-Laven** [reply]
@viceice, is there some benefit to using `packageRules` as opposed to [the `lockFileMaintenance` key](https://github.com/ScribeMD/.github/blob/5a8d5a3434d0c35d9a5ba896cae74ef03022624a/default.json#L26-L29)? I think we would need a bogus schedule like "on the 29th day of February in 1970" either way in order to disable the schedule?

@rarkins, is there any way to delete the `renovate/lock-file-maintenance` branch in that case? I confirmed that the branch exists on the forks of each of our repositories:

- https://github.com/renovate-bot/template-repo
- https://github.com/renovate-bot/.github-6
- https://github.com/renovate-bot/rootless-docker
- https://github.com/renovate-bot/docker-cache
- https://github.com/renovate-bot/pre-commit-action
- https://github.com/renovate-bot/pre-commit-hooks
- https://github.com/renovate-bot/slack-templates

**Reply by rarkins** [reply]
If you view the closed PR, does the github UI give you the option to delete the source branch? The bot sets "maintainers can modify" for such PRs currently but I'm not sure if that allows you to also *delete*.

**Reply by Kurt-von-Laven** [reply]
Ah, good thought, and I know the button you are referring to, but sadly no, there's no such option in this case.

**Reply by viceice** [reply]
@Kurt-von-Laven yes, packages rules are applied after the object keys.

diable schedule: `"schedule": []` it's same as `at any time`

**Reply by Kurt-von-Laven** [reply]
> yes, packages rules are applied after the object keys.

Are you saying that the default lock file maintenance schedule is effectively specified as an object key, so it is preferable to use a package rule to ensure that it is overridden? Otherwise I'm not sure what settings would potentially be in conflict.

When last I tried setting `"at any time"` for lock file maintenance via the object key, Renovate began opening lock file maintenance pull requests incessantly.

**Reply by viceice** [reply]
yes, thats expected. whithout schedule and without dashboard approval renovate will create a lockfile maintenance pr when there is a lockfile change. renovate will check this on every run. so you need to use a schedule or dashboard approval to reduce noise.

you can also limit branch updates to schedule via `updateNotScheduled`[^1]

[^1]: https://docs.renovatebot.com/configuration-options/#updatenotscheduled

**Reply by Kurt-von-Laven** [reply]
To clarify, I am describing the behavior with Dependency Dashboard approval already enabled and a lock file maintenance schedule of `"at any time"`.

**Reply by rarkins** [reply]
Setting prCreation=approval may work

**Reply by Kurt-von-Laven** [reply]
I gave that a shot, and it prevented lock file maintenance PRs from being created when we check the appropriate box in the Dependency Dashboard when lock file maintenance is in the "Awaiting Schedule" state. I confirmed that there is lock file maintenance to perform. Notably, using `"prCreation" = "approval"` did not prevent lock file maintenance PRs from being created by checking the box when the branch was already in the "PR Creation Approval Required" state. Since we have Dependency Dashboard approval enabled, it also means that, for all branches, we need to check the box once to move them from the "Pending Approval" or "Awaiting Schedule" state to the "PR Creation Approval Required" and then again to actually create the PR.

**Comment by rarkins** [comment]
This problem seems like it could be specific to forking renovate, because part of the cause is that branches stick around and aren't auto deleted.

I think if there was a possibility to autodelete merged PR branches then it might behave as expected.

Otherwise, Renovate is faced with the following challenge: it sees the lock file maintenance branch exists, so what should it do with it? e.g. should it de-conflict and update it? If so then you may end up with non stop lock file maintenance PRs because dependencies change frequently.

If it is to _delete_ the existing branch, we need logic to ensure it won't malfunction and delete branches which are meant to be there. I suggest one heuristic would be if the branch exists but a PR doesn't.

**Reply by Kurt-von-Laven** [reply]
Our repositories are configured to automatically delete merged PR branches, but I assume the setting does not apply to forks. If Forking Renovate runs and detects that one of its branches has been merged with the contents currently on that branch (or, essentially equivalently, that the branch contains no unique contents after a rebase), then it can safely delete the branch. Such logic would have to precede any updates to the branch, and while I don't see any harm in checking that no PR is open, I don't think it would be necessary if one were to check that the branch rebased clean. Renovate could decide as appropriate whether to force-push the rebased (and potentially subsequently updated) version of the branch if it didn't delete it.

**Reply by Kurt-von-Laven** [reply]
If I'm not mistaken, Forking Renovate already automatically runs a job when one of its PRs is merged, so that seems like an opportunity to delete the branch, essentially implementing auto-deletion of merged PR branches.
